pkg/istio/mesh_config.go

```diff
@@ -17,12 +17,115 @@ import (
 	maistrav2 "github.com/kuadrant/kuadrant-operator/api/external/maistra/v2"
 )
 
-// The structs below implement the interface defined in pkg/common/mesh_config.go `ConfigWrapper`
-// type ConfigWrapper interface {
-//		GetConfigObject() client.Object
-//		GetMeshConfig() (*istiomeshv1alpha1.MeshConfig, error)
-//		SetMeshConfig(*istiomeshv1alpha1.MeshConfig) error
-// }
+const (
+	ExtAuthorizerName = "kuadrant-authorization"
+)
+
+type authorizer interface {
+	GetExtensionProvider() *istiomeshv1alpha1.MeshConfig_ExtensionProvider
+}
+
+type ConfigWrapper interface {
+	GetConfigObject() client.Object
+	GetMeshConfig() (*istiomeshv1alpha1.MeshConfig, error)
+	SetMeshConfig(*istiomeshv1alpha1.MeshConfig) error
+}
+
+type KuadrantAuthorizer struct {
+	extensionProvider *istiomeshv1alpha1.MeshConfig_ExtensionProvider
+}
+
+// NewKuadrantAuthorizer Creates a new KuadrantAuthorizer
+func NewKuadrantAuthorizer(namespace string) *KuadrantAuthorizer {
+	return &KuadrantAuthorizer{
+		extensionProvider: createKuadrantAuthorizer(namespace),
+	}
+}
+
+// GetExtensionProvider Returns the Istio MeshConfig ExtensionProvider for Kuadrant
+func (k *KuadrantAuthorizer) GetExtensionProvider() *istiomeshv1alpha1.MeshConfig_ExtensionProvider {
+	return k.extensionProvider
+}
```

Comment:
I thought we were keeping these interfaces and functions at the abstract level. Does this mean that having a Kuadrant Authorizer became a Istio-specific thing?

Reply:
yes, it is an istio specific thing.

Reply:
I don't follow. We won't deploy Authorino when integrating with Envoy Gateway?

Reply:
Yes, we will. But this "kuadrant authorizer" is not deploying authorino. The kuadrant controller is. The "kuadrant authorizer" is to register external auth in istio control plane. AFAIK, EnvoyGateway does not require any registration step in the control plane of EG. @adam-cattermole can confirm.

Reply:
I understand, @eguzki. My point is that this abstraction ("Kuadrant Authorizer config") has not been as much of an abstraction as it could be or [originally intended](https://github.com/Kuadrant/kuadrant-operator/pull/192#discussion_r1219630839).

At the level of the Kuadrant CR/Kuadrant deployment, _configuring the Kuadrant Authorizer_ means ensuring that the Kuadrant deployment has all the configuration it needs for ext_authz requests to be sent to an instance of Authorino in the data plane, when handling traffic for the gateways in the scope of this Kuadrant deployment and it's applicable.

Because so far we only support Istio-based Gateway API controllers, it's easy overlook the difference, but I'm arguing for the decoupling of the abstraction that ensures the external authorizer is configured for a given Gateway API controller from the implementation details that get it done.

IWO, these interfaces belong to the Kuadrant controller and should be fixed to not know as much as they do today about Istio IMHO. In fact, they should not know anything at all about Istio. Only the Istio-specific implementations should know that the way to get the external authorizer configured (in their case) is by patching a "mesh config." Perhaps for other Gateway API controllers (e.g. EG), configuring the external authorizer at the level of the Kuadrant CR means something else; it could even be a no-op – but even that should be hidden behind the abstraction IMO.

Now that we are decoupling the controller from Istio, it looked like the right opportunity to fix this.

Reply:
That abstraction is not meant to decouple from gateway API provider. It is totally coupled to Istio. 

```go
type ConfigWrapper interface {
	GetConfigObject() client.Object
	GetMeshConfig() (*istiomeshv1alpha1.MeshConfig, error)
	SetMeshConfig(*istiomeshv1alpha1.MeshConfig) error
}

type KuadrantAuthorizer struct {
	extensionProvider *istiomeshv1alpha1.MeshConfig_ExtensionProvider
}
```

I think the intended design wanted to abstract from the different "flavors" of Istio. But actually, IMHO, there is no abstraction actually. It is something like: "if istioOperator -> do this; else if istio configmap -> do that; else if ossm -> do other thing; else panic". 

Anyway, I get your point and I will try something I had in mind.

Reply:
Prefer to stab it in separate PR to unblock this one? It's all refactoring in the end.